### PR TITLE
Optimize `inline_memcpy` by `avx2`

### PR DESCRIPTION
**Pull request by solotzg**
Signed-off-by: Zhigao Tong <tongzhigao@pingcap.com>

### What problem does this PR solve?

Issue Number: ref #6233

Problem Summary:

### What is changed and how it works?

- optimize [inline_memcpy](https://github.com/pingcap/tiflash/blob/dfac6a50b09711d12a3b6909126782d1b8ed363d/libs/libcommon/include/common/memcpy.h#L25) with avx2 flags
- For implementation about `avx2_inline_memcpy`, optimize small size memory
- add cmake functions `check_then_add_sources_compile_flag` to check then add compile flags for source files.
- add avx2 flag to source files `ColumnString.cpp` and `ColumnsCommon.cpp`

### Benchmark

```
Run on (40 X 2386.24 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x20)
  L1 Instruction 32 KiB (x20)
  L2 Unified 256 KiB (x20)
  L3 Unified 25600 KiB (x2)
Load Average: 6.72, 6.65, 6.07
-------------------------------------------------------------------------------------------------------------------
Benchmark                                                                         Time             CPU   Iterations
-------------------------------------------------------------------------------------------------------------------
MemUtilsCopy_1_20_3_true_20000/stl_mempy/iterations:500                      446194 ns       446197 ns          500
MemUtilsCopy_1_20_3_true_20000/inline_clickhouse_memcpy/iterations:500       140571 ns       140528 ns          500
MemUtilsCopy_1_20_3_true_20000/sse2_memcpy/iterations:500                    150496 ns       150321 ns          500
MemUtilsCopy_1_20_3_true_20000/avx2_memcpy/iterations:500                    137616 ns       137586 ns          500
MemUtilsCopy_1_20_3_true_20000/folly_memcpy/iterations:500                   181997 ns       182004 ns          500
MemUtilsCopy_1_40_3_true_20000/stl_mempy/iterations:500                      456534 ns       456549 ns          500
MemUtilsCopy_1_40_3_true_20000/inline_clickhouse_memcpy/iterations:500       176257 ns       176264 ns          500
MemUtilsCopy_1_40_3_true_20000/sse2_memcpy/iterations:500                    178094 ns       178086 ns          500
MemUtilsCopy_1_40_3_true_20000/avx2_memcpy/iterations:500                    165250 ns       165256 ns          500
MemUtilsCopy_1_40_3_true_20000/folly_memcpy/iterations:500                   186537 ns       186544 ns          500
MemUtilsCopy_1_80_3_true_20000/stl_mempy/iterations:500                      476534 ns       476510 ns          500
MemUtilsCopy_1_80_3_true_20000/inline_clickhouse_memcpy/iterations:500       238987 ns       238996 ns          500
MemUtilsCopy_1_80_3_true_20000/sse2_memcpy/iterations:500                    302893 ns       302887 ns          500
MemUtilsCopy_1_80_3_true_20000/avx2_memcpy/iterations:500                    231339 ns       231348 ns          500
MemUtilsCopy_1_80_3_true_20000/folly_memcpy/iterations:500                   200658 ns       200665 ns          500
MemUtilsCopy_1_200_3_true_20000/stl_mempy/iterations:500                     660466 ns       660356 ns          500
MemUtilsCopy_1_200_3_true_20000/inline_clickhouse_memcpy/iterations:500      440048 ns       439948 ns          500
MemUtilsCopy_1_200_3_true_20000/sse2_memcpy/iterations:500                   427743 ns       427739 ns          500
MemUtilsCopy_1_200_3_true_20000/avx2_memcpy/iterations:500                   198850 ns       198876 ns          500
MemUtilsCopy_1_200_3_true_20000/folly_memcpy/iterations:500                  243985 ns       243973 ns          500
MemUtilsCopy_1_2000_3_true_20000/stl_mempy/iterations:500                   1535333 ns      1535242 ns          500
MemUtilsCopy_1_2000_3_true_20000/inline_clickhouse_memcpy/iterations:500     847673 ns       847547 ns          500
MemUtilsCopy_1_2000_3_true_20000/sse2_memcpy/iterations:500                  793617 ns       793608 ns          500
MemUtilsCopy_1_2000_3_true_20000/avx2_memcpy/iterations:500                  682924 ns       682915 ns          500
MemUtilsCopy_1_2000_3_true_20000/folly_memcpy/iterations:500                 662669 ns       662652 ns          500
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: libs/libcommon/include/common/avx2_memcpy.h [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        if (size <= block32_size)
+        {
+            MEM_CP_END(block32_size);
+            return ret;
+        }
+    }
+
+    assert(size > block32_size);
+
+    if (unlikely(size <= 2 * block32_size))
```
Comment (Lloyd-Pottiger) [inline libs/libcommon/include/common/avx2_memcpy.h:137]:
```suggestion
    if (unlikely(size <= (block32_size << 1)))
```

use shift operator to get better performance?

Comment (solotzg) [reply libs/libcommon/include/common/avx2_memcpy.h:137]: Modern compiler itself can optimize such simple case.

Comment (Lloyd-Pottiger) [reply libs/libcommon/include/common/avx2_memcpy.h:137]: ok

**Review by Lloyd-Pottiger — APPROVED** [review]
LGTM
Will we do similar optimization for `memcpy`?

**Comment by solotzg** [comment]
> LGTM Will we do similar optimization for `memcpy`?

The default value of `USE_INTERNAL_MEMCPY` is `ON` under x86-64 platform.

https://github.com/pingcap/tiflash/blob/40ad20c9439b18221297ed2b87d9d7421d17b174/libs/libmemcpy/CMakeLists.txt#L31-L42

The global memcpy function will be overrided by implementation here.
https://github.com/pingcap/tiflash/blob/dfac6a50b09711d12a3b6909126782d1b8ed363d/libs/libmemcpy/memcpy.cpp#L23

**Comment by Lloyd-Pottiger** [comment]
sorry, typo…… `memcmp`

**Comment by solotzg** [comment]
> LGTM Will we do similar optimization for `memcpy`?
> sorry, typo…… `memcmp`

By now, the comparison func of `StringRef` has been replaced with ours for most known cases(only need 3 options positive,negative,zero).
But for memcmp, I'm not sure whether other modules rely on its return value.

**Comment by solotzg** [comment]
> > LGTM Will we do similar optimization for `memcpy`?
> > sorry, typo…… `memcmp`
> 
> By now, the comparison func of `StringRef` has been replaced with ours for most known cases(only need 3 options positive,negative,zero). But for memcmp, I'm not sure whether other modules rely on its return value.

Maybe that's not a problem and I'm overthinking.

**Comment by Lloyd-Pottiger** [comment]
> > LGTM Will we do similar optimization for `memcpy`?
> > sorry, typo…… `memcmp`
> 
> By now, the comparison func of `StringRef` has been replaced with ours for most known cases(only need 3 options positive,negative,zero). But for memcmp, I'm not sure whether other modules rely on its return value.

so, the best practice for now is to replace `memcmp` with `mem_utils::IsStrViewEqual` / `mem_utils::CompareStrView`?

**Comment by solotzg** [comment]
> so, the best practice for now is to replace `memcmp` with `mem_utils::IsStrViewEqual` / `mem_utils::CompareStrView`?

I think so. Using internal code can help compiler to inline it.

**Review by zanmato1984 — APPROVED** [review]
LGTM

**Comment by Lloyd-Pottiger** [comment]
> > so, the best practice for now is to replace `memcmp` with `mem_utils::IsStrViewEqual` / `mem_utils::CompareStrView`?
> 
> I think so. Using internal code can help compiler to inline it.

get it, thank you

**Pull request merged**